src/core/ext/transport/cronet/transport/cronet_transport.c

```
@@ -0,0 +1,583 @@
+  uint8_t *p = s->write_buffer;
+  // Append 5 byte header
+  *p++ = 0;
+  *p++ = (uint8_t)(length >> 24);
+  *p++ = (uint8_t)(length >> 16);
+  *p++ = (uint8_t)(length >> 8);
+  *p++ = (uint8_t)(length);
+  // append actual data
+  memcpy(p, raw_data, length);
+}
+
+void do_write(stream_obj *s) {
+  gpr_slice_buffer *sb = &s->write_slicebuffer;
+  GPR_ASSERT(sb->count <= 1);
+  if (sb->count > 0) {
+    create_grpc_frame(s);
+    GRPC_CRONET_TRACE(GPR_DEBUG, "W: cronet_bidirectional_stream_write");
+    cronet_bidirectional_stream_write(s->cbs, s->write_buffer,
+                                      (int)s->write_buffer_size, false);
+  }
+}
+
+//
+void next_send_step(stream_obj *s) {
+  switch (cronet_send_state) {
+    case CRONET_SEND_IDLE:
+      GPR_ASSERT(
+          s->cbs);  // cronet_bidirectional_stream is not initialized yet.
+      cronet_send_state = CRONET_REQ_STARTED;
+      GRPC_CRONET_TRACE(GPR_DEBUG, "cronet_bidirectional_stream_start to %s",
+                        s->url);
+      cronet_bidirectional_stream_start(s->cbs, s->url, 0, "POST",
+                                        &s->header_array, false);
+      break;
+    case CRONET_SEND_HEADER:
+      do_write(s);
+      cronet_send_state = CRONET_WRITE;
+      break;
+    case CRONET_WRITE_COMPLETED:
+      do_write(s);
+      break;
+    default:
+      GPR_ASSERT(0);
+      break;
+  }
+}
+
+void create_url(const char *path, const char *host, stream_obj *s) {
+  const char prefix[] = "https://";
+  s->url = gpr_malloc(strlen(prefix) + strlen(host) + strlen(path) + 1);
+  strcpy(s->url, prefix);
+  strcat(s->url, host);
+  strcat(s->url, path);
+}
+
+static void convert_metadata_to_cronet_headers(grpc_linked_mdelem *head,
+                                               const char *host,
+                                               stream_obj *s) {
+  grpc_linked_mdelem *curr = head;
+  while (s->num_headers < MAX_HDRS) {
```

Silently discarding seems to be a bad option